### PR TITLE
Chunking

### PR DESCRIPTION
**Pull request by Hoijof**

### CODE REVIEW

File: src/game/Circle.ts [hunk 1]
```diff
@@ -74,7 +106,7 @@ export default class Circle extends GameObject {
     static parse(context, data) {
         const obj = data;
 
-        return new Circle(context, obj[1], obj[2], obj[3], obj[4], obj[5], obj[6]);
+        return new Circle(context, obj[1], obj[2], obj[3], obj[4], obj[5], obj[6], obj[7]);
```
Comment (amatiasq) [inline src/game/Circle.ts:109]: bruh

```js
return new Circle(context, ...obj.slice(1, 7));
```

File: src/game/GameObject.ts [hunk 2]
```diff
@@ -1,24 +1,47 @@
-import { Camera } from "../types/types";
+import { Camera, Types } from "../types/types";
+
+let lastId = 0;
+
+function toFixedIfNecessary(value, dp) {
+    return +parseFloat(value).toFixed(dp);
+}
```
Comment (amatiasq) [inline src/game/GameObject.ts:7]: dupl

File: src/game/GameObject.ts [hunk 3]
```diff
@@ -1,24 +1,47 @@
...
         this.vy = vy;
         this.mass = mass;
 
         this.isColliding = false;
+        this.dirtyList = new Set();
     }
 
     draw(camera: Camera) { }
+    update(delta: number): Array<any> { return null; }
+    generateUpdateActions(): Array<any> { return null }
```
Comment (amatiasq) [inline src/game/GameObject.ts:40]: Use TS strict nulls, changes your life

File: src/game/GameWorld.ts [hunk 4]
```diff
@@ -61,7 +70,68 @@ export default class GameWorld {
...
+        }
+
+        if (cy >= MAP_SIZE) {
+            cy = MAP_SIZE - 10;
         }
+
+        return {
+            x: Math.floor(cx / this.chunkSize),
+            y: Math.floor(cy / this.chunkSize)
+        };
```
Comment (amatiasq) [inline src/game/GameWorld.ts:134]: Esto con `Vector` no pasaba

File: src/game/GameWorld.ts [hunk 5]
```diff
@@ -73,7 +143,10 @@ export default class GameWorld {
 
     createWorld() {
         for (let i = 0; i < INITIAL_ENTITIES; i++) {
-            this.gameObjects.push(Circle.createRandom(this.context, this.world));
+            const circle: Circle = Circle.createRandom(this.context, this.world);
```
Comment (amatiasq) [inline src/game/GameWorld.ts:146]: `circle: Circle = Circle`
why `createRandom` doesn't return `Circle` already?

File: src/game/GameWorld.ts [hunk 6]
```diff
@@ -99,25 +172,53 @@ export default class GameWorld {
...
+            this.addToMap(go);
         }
 
-        let collisionTime;
         if (this.env === RENDER_ENV.NODE) {
-            collisionTime = performance.now();
+            global.reporter.report('updateTime', performance.now() - updateTime);
+        }
+
+        const collisionTime = this.env === RENDER_ENV.NODE ? performance.now() : null;
```
Comment (amatiasq) [inline src/game/GameWorld.ts:195]: Second appearance in 20 lines, encapsulate in function

File: src/game/GameWorld.ts [hunk 7]
```diff
@@ -99,25 +172,53 @@ export default class GameWorld {
...
+        if (USE_CHUNK_UPDATE) {
+            this.detectCollisionsChunks();
+        } else {
+            this.detectCollisions(this.gameObjects);
         }
 
-        this.detectCollisions();
         this.detectEdgeCollisions();
 
+        if (USE_PARTIAL_UPDATE && this.env === RENDER_ENV.NODE) {
```
Comment (amatiasq) [inline src/game/GameWorld.ts:205]: Second appearance in 20 lines, encapsulate in getter

File: src/game/GameWorld.ts [hunk 8]
```diff
@@ -209,21 +321,62 @@ export default class GameWorld {
...
+            * 
+            *    X   X   O
+            *    
+            *    X   C   O
+            *  
+            *    O   O   O
+             */
+            for (let i = ci - 1; i <= ci + 1; ++i) {
+                for (let j = cj - 1; j <= cj + 1; ++j) {
+                    if (i < 0 || j < 0 || i >= CHUNK_COUNT || j >= CHUNK_COUNT || (i < ci && j < cj) || (i === ci && j < cj) || (i < ci && j === cj)) {
```
Comment (amatiasq) [inline src/game/GameWorld.ts:353]: unreadable, decompose into vars bro

File: src/game/GameWorld.ts [hunk 9]
```diff
@@ -209,21 +321,62 @@ export default class GameWorld {
...
+                        continue;
+                    }
+
+                    entities.push(...this.chunkMap[i][j]);
+                }
+            }
+
+            if (this.env === RENDER_ENV.NODE) {
```
Comment (amatiasq) [inline src/game/GameWorld.ts:361]:
```js
get IS_NODE() {
  return this.env === RENDER_ENV.NODE
}
```

File: src/game/GameWorld.ts [hunk 10]
```diff
@@ -288,20 +441,74 @@ export default class GameWorld {
         return circleDistance <= ((r1 + r2) * (r1 + r2))
     }
 
-    serialize(): Array<GameObject> {
+    serialize(): any {
+        if (USE_PARTIAL_UPDATE && this.env === RENDER_ENV.NODE) {
```
Comment (amatiasq) [inline src/game/GameWorld.ts:445]: third time

File: src/game/GameWorld.ts [hunk 11]
```diff
@@ -288,20 +441,74 @@ export default class GameWorld {
...
+            this.generateChunkMap();
+            this.gameObjects.forEach((object: GameObject) => this.addToMap(object));
+
+            return;
+        }
+        if (data[StateTypes.fullState]) {
+            this.clearGameObjects();
+            data[StateTypes.fullState].forEach(fullState => {
+                switch (fullState[0]) {
+                    case Types.CIRCLE:
```
Comment (amatiasq) [inline src/game/GameWorld.ts:469]: BERI USEFUL SWITCH

Comment (amatiasq) [reply src/game/GameWorld.ts:469]: encapsulate

```js
const deserializer = {
  [Types.CIRCLE](context, state) {
    Circle.createFromCreateAction(context, state);
  }
};
```

File: src/game/GameWorld.ts [hunk 12]
```diff
@@ -288,20 +441,74 @@ export default class GameWorld {
...
+                        this.gameObjects.push(Circle.createFromCreateAction(this.context, fullState));
+                        this.idToIndex[fullState[1]] = this.gameObjects.length - 1;
+                }
+            });
+        }
+
+        if (data[StateTypes.createAction]) {
+            data[StateTypes.createAction].forEach(createAction => {
+                switch (createAction[0]) {
+                    case Types.CIRCLE:
```
Comment (amatiasq) [inline src/game/GameWorld.ts:479]: Idem, and reuse here

File: src/game/StateUpdate.ts [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+      this.state[type] = [];
+    }
+
+    this.state[type].push(content);
+  }
+
+  toJSON() {
+    return this.state;
+  }
+}
```
Comment (amatiasq) [inline src/game/StateUpdate.ts:27]: no EOL
dafuk man

File: src/server/websocket/client.ts [hunk 14]
```diff
@@ -31,7 +31,7 @@ export default class HClient {
     this.server.parseMessage(decodedMessage, this);
   }
 
-  send(data: string, type: HMessageTypes = HMessageTypes.string) {
+  send(data: any, type: HMessageTypes = HMessageTypes.string) {
```
Comment (amatiasq) [inline src/server/websocket/client.ts:34]: cartabuena

File: src/server/websocket/server.ts [hunk 15]
```diff
@@ -85,11 +85,11 @@ export default class HServer {
     }
   }
 
-  broadcast(message: string, type: HMessageTypes) {
+  broadcast(message: string, type: HMessageTypes, directSend: boolean = false) {
```
Comment (amatiasq) [inline src/server/websocket/server.ts:88]: `: boolean` is redundant here

**Review by amatiasq — APPROVED** [review]
LGTM, deploy to PROD

**Comment by amatiasq** [comment]
Long-lived feature branches are the root of all evil

File: src/game/Circle.ts [hunk 16]
```diff
@@ -13,39 +13,50 @@ function toFixedIfNecessary(value, dp) {
...
 
-    update(delta) {
+    basicUpdate(basicUpdate: number[]) {
+        this.x = basicUpdate[1];
+        this.y = basicUpdate[2];
+    }
+
+    updateAction(updateAction: number[]) {
+        for (let i = 1; i < updateAction.length; i += 2) {
+            this[this.indexToKey[updateAction[i]]] = updateAction[i + 1];
```
[the remark this answers is not in the document]
Comment (Hoijof) [reply src/game/Circle.ts:53]: Update actions follow the following structure:

[id, key, value, key2, value2, key3, value3] They have all the discretional updates for a given entity, subnormal.

File: src/game/Circle.ts [hunk 17]
```diff
@@ -13,39 +13,50 @@ function toFixedIfNecessary(value, dp) {
...
+    keyToIndex: Object;
+
+    constructor(context, id, x, y, r, mass, vx = getRandomInt(-20, 20), vy = getRandomInt(-20, 20)) {
+        super(context, id, x, y, r, vx, vy, mass);
+
+        this.type = Types.CIRCLE;
+        //                  0       1     2    3     4      5     6
+        this.indexToKey = ['type', 'id', 'x', 'y', 'mass', 'vx', 'vy'];
+
+        this.keyToIndex = this.indexToKey.reduce((acc, value, index) => {
```
[the remark this answers is not in the document]
Comment (Hoijof) [reply src/game/Circle.ts:30]: Why? That's just slower, you need to create an array for every iteration :D  and I think it's still hard to read after this change

File: src/game/Circle.ts [hunk 18]
```diff
@@ -55,11 +66,32 @@ export default class Circle extends GameObject {
 
         // Convert to degrees
         //let degrees = 180 * radians / Math.PI;
+
+        if (this.vx === 0 && this.vy === 0) {
```
[the remark this answers is not in the document]
Comment (Hoijof) [reply src/game/Circle.ts:70]: We don't do that here

File: src/game/GameObject.ts (hunk 3 above)
Comment (Hoijof) [reply src/game/GameObject.ts:40]: everything can be null in any real language.

File: src/game/Circle.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (amatiasq) [reply src/game/Circle.ts:70]: cartabuena

File: src/game/Circle.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (amatiasq) [reply src/game/Circle.ts:30]: tu cara es más difícil de leer

**Pull request merged**